examples/07 Code Generation.ipynb

```diff
@@ -295,99 +285,90 @@
+    "   }\n",
+    "   repeated class2 class_name = 3;\n",
+    "}\n",
     "\n",
-    ">    datatype class1_name1 = 1;\n",
-    "\n",
-    ">    dataype class1_name2 = 2;\n",
-    "\n",
-    ">    message class2 {\n",
-    "\n",
-    "> >  datatype class2_name1 = 1;\n",
-    "\n",
-    "> >  datatype class2_name2 = 2;\n",
-    "\n",
-    "> >  datatype class2_name3 = 3;\n",
-    "\n",
-    " >   }\n",
-    "\n",
-    ">    repeated class2 class_name = 3;\n",
-    "    \n",
-    "}\n"
+    "```\n"
    ]
   },
   {
    "cell_type": "code",
-   "execution_count": 9,
+   "execution_count": 28,
    "metadata": {},
    "outputs": [],
    "source": [
-    "def class_to_proto(cls, current_classes=[], indent=\"\\t\"):\n",
-    "    if not current_classes:\n",
-    "        text = ['syntax = \"proto3\";\\n']\n",
-    "    text = [f\"{indent[:-1]}message  {cls.name} {{\"]\n",
-    "    current_classes.append([cls])\n",
+    "def class_to_proto(cls, current_classes=None, indent=\"\"):\n",
+    "    if current_classes is None:\n",
+    "        current_classes = [cls]\n",
+    "        lines = ['syntax = \"proto3\";\\n']\n",
+    "        indent += \" \"*4\n",
+    "    lines = [f\"{indent[:-1]}message  {cls.name} {{\"]\n",
     "\n",
-    "    for counter, prop in enumerate(cls.properties):\n",
-    "        multiplicity = (\"\", \"[]\")[prop.max_card.value > 1.0]\n",
+    "    for counter, prop in enumerate(cls.properties, start=1):\n",
+    "        multiplicity = (\"\", \"[]\")[prop.max_card.value > 1]\n",
     "        if (\n",
     "            isinstance(\n",
     "                prop.type, capellambse.model.crosslayer.information.Class\n",
     "            )\n",
     "            and prop.type not in current_classes\n",
     "        ):\n",
-    "            current_classes.append([prop.type])\n",
+    "            current_classes.append(prop.type)\n",
     "            nested_text = class_to_proto(\n",
-    "                prop.type, current_classes, indent + \"\\t\"\n",
+    "                prop.type, current_classes, indent + \" \"*4\n",
     "            )\n",
-    "            text.append(nested_text)\n",
-    "            text.append(\n",
-    "                f\"{indent}repeated {prop.type.name}{multiplicity} {prop.name} = {counter+1};\"\n",
+    "            lines.append(nested_text)\n",
+    "            lines.append(\n",
+    "                f\"{indent}repeated {prop.type.name}{multiplicity} {prop.name} = {counter};\"\n",
     "            )\n",
     "        else:\n",
-    "            text.append(\n",
-    "                f\"{indent}{prop.type.name}{multiplicity} {prop.name} = {counter+1};\"\n",
+    "            lines.append(\n",
+    "                f\"{indent}{prop.type.name}{multiplicity} {prop.name} = {counter};\"\n",
     "            )\n",
-    "    text[-1] = text[-1] + f\"\\n{indent[:-1]}}}\"\n",
-    "    text = \"\\n\".join(text)\n",
+    "    lines[-1] = lines[-1] + f\"\\n{indent[:-1]}}}\"\n",
```

Comment:
You can use `+=` here. But really, you should rather use `lines.append()` instead.

Reply:
Okay, I use `+=`, because `lines[-1] is a `str` object which has no attribute `append` :)